### PR TITLE
Foundation code for Message Groups

### PR DESCRIPTION
**Pull request by telyn**
Hey again :)

I got started working on #1183 and I thought I'd get this PR made sooner rather than later so you can take a sneak peek if you want - I've got a decent amount of the foundational work done towards it without (I believe) causing any breakage.

**Mostly** though I want to ask the question: How would you like to organise getting my changes PR'd into danger? All in one massive PR when I'm all done, or piece-by-piece?

The summary of the changes I've made are that:
* The `MessageAggregator` class has been added without any methods.
* `MessageAggregator.aggregate` (which does not exist) will be called from `Dangerfile#post_results`, and it will be created if a request_source has a `#update_pr_by_line!` method (which none currently do) (I am very happy to change `#update_pr_by_line!`'s name - it's the best I could come up with at the time!)
* The `MessageGroup` type has been added and has a fairly decent sketch of its methods.
* Violation now has a `#type`, which defaults to `:warning`, can be set from calls to `.new`, and all calls to `.new` inside this codebase pass it - so `fail`,`warn`, `message` all create `Violations` with the correct `type#`
* Markdown now also has `#type` - which returns `:markdown`.
* Violation and Markdown now have a thoroughly-tested `#<=>` method each. in preparation for the sorting code being added to `MessageGroup`.
* Violation and Markdown now inherit from a base type to keep the code a bit DRY-er. In retrospect perhaps the base class should've had a list of types and their orderings - but I only made the base after I'd started on the implementation for sorting them both 😅

A changelog to that effect will be along shortly :)

P.s. is it cool if I add [Fuubar](https://github.com/thekompanee/fuubar) and fiddle with the guardfile so it doesn't run rubocop if the rspec tests fail?

### CODE REVIEW

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/danger/danger/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,17 +9,21 @@
...
+* Fix incorrect method call in Bitbucket Cloud arising from my own changes in 6.2.1 [@telyn][@telyn] [#1184](https://github.com/danger/danger/pull/1184)
 
 ## 6.2.1
 
 * Fix Bitbucket Pipelines documentation
 * Update Ruby-Git to 1.6.0 to fix output encoding
-* Add inline-comments support to Bitbucket Cloud [@telyn](https://github.com/telyn) [#1182](https://github.com/danger/danger/pull/1182)
-* Make emojis display better on Bitbucket Cloud [@telyn](https://github.com/telyn) [#1182](https://github.com/danger/danger/pull/1182)
+* Add inline-comments support to Bitbucket Cloud [@telyn][@telyn] [#1182](https://github.com/danger/danger/pull/1182)
+* Make emojis display better on Bitbucket Cloud [@telyn][@telyn] [#1182](https://github.com/danger/danger/pull/1182)
```
Comment (orta) [inline CHANGELOG.md]: If you leave this blank, then markdown will assume it is the same as the text on the left -`[@telyn][]` === `[@telyn][@telyn]`

File: .rubocop.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Defaults can be found here: https://github.com/bbatsov/rubocop/blob/master/config/default.yml
 
 AllCops:
-  TargetRubyVersion: 2.0
+  TargetRubyVersion: 2.3
```
Comment (orta) [inline .rubocop.yml]: Seems legit

File: CHANGELOG.md (hunk 1 above)
Comment (telyn) [reply CHANGELOG.md]: oh, neat, didn't know that 😁

File: .rubocop.yml (hunk 2 above)
Comment (telyn) [reply .rubocop.yml]: I needed it to stop rubocop complaining, and the gemspec says 2.3 is the minimum supported version anyway 😉

**Comment by orta** [comment]
I'm happy for you to add deverloper-y dependencies, I only get cagey about dependencies which danger consumers end up with.

WRT to how you want to get it in, I prefer that master is always deployable - this looks backwards compat, but if it weren't we could just scope it like:

```ruby
if env.request_source.respond_to?(:update_pr_by_line!) && process.env.USE_NEW_THING
``` 

So, feels good to me

**Comment by telyn** [comment]
Cool - I'm gonna deal with your suggestion on the markdown file, dogfood it more at work to ensure that I've not broken anything, then un-WIP this :)

**Pull request merged**